src/coreclr/vm/syncblk.cpp

```diff
@@ -2237,10 +2237,14 @@ SyncBlock *ObjHeader::GetSyncBlock()
 
                 LEAVE_SPIN_LOCK(this);
             }
-            // SyncBlockCache::LockHolder goes out of scope here
         }
+        // SyncBlockCache::LockHolder goes out of scope here
     }
 
+    // Fire the LockCreated event last, such that it is outside the FAULT_FORBID region above and outside the SyncBlockCache
+    // lock. The eventing system may need to allocate.
+    syncBlock->FireLockCreatedEvent();
```

Comment:
Does this guarantee that we only fire this event when a lock is created?  There are other paths that create syncblocks, right?

Reply:
I didn't see any other paths that create sync blocks

Reply:
From what I can see, there are others.  Here's an example: https://github.com/dotnet/runtime/blob/main/src/coreclr/vm/object.cpp#L987

Reply:
It looks like this path also goes through `GetSyncBlock()`, which is modified here, so it would also emit the event

Reply:
Right.  There are others too, so I don't think this is the right place to emit the event, as it will be potentially much more verbose (and duplicative) than we want.

Reply:
The `AwareLock` is created when the `SyncBlock` is created, the monitor is an instance field. Even if the object hasn't been used for locks yet, the `AwareLock` would be created at this point.

Reply:
That makes sense, and I think it makes sense at that point to log the event.  I just don't think we should log the event each time the syncblock is acquired, since that could happen much more often than the first time that it is created.

Reply:
> I just don't think we should log the event each time the syncblock is acquired, since that could happen much more often than the first time that it is created.

I thought the sync block would be created once per object and no more? I figured the event would be emitted when the sync block is created for an object, and the sync block may be reused for different purposes thereafter. Sync block creation could be frequent, but probably less frequent than lock contention.

Reply:
> I thought the sync block would be created once per object and no more? I

That's only true for precious syncblocks. Syncblocks that are not marked as precious can be reclaimed and re-created multiple times.

Reply:
> That's only true for precious syncblocks. Syncblocks that are not marked as precious can be reclaimed and re-created multiple times.

Wouldn't that mean that each time a sync block is recreated for an object, that it may have a different pointer to the `AwareLock` and hence warrant emitting the event?

Reply:
Yes, that's correct.

I think that Brian's concern is that we will emit this event for all SyncBlocks even the ones that are not ever used for locks. SyncBlocks are created for number of different reasons (the name "syncblock" is misleading). For example, interop creates them a lot for COM objects and marshalled delegates.

Reply:
Yes.  What do you think about having the `AwareLock` log that it was created lazily - on first acquisition?

Reply:
I see. Yes, it could probably be emitted lazily. Taking a step back, is the lock ID really necessary? The stack trace of the contention event would probably be sufficient to identify the lock. I figure the `LockCreated` event's stack trace would provide similar information. A benefit may be to have only a few `LockCreated` events and avoid taking stack traces for contention events, but at the cost of missing those created events in some cases. Is there any other benefit?